projects/asn1-proto/asn1_proto_to_der.cc

```diff
@@ -0,0 +1,156 @@
+    // to encode the length.
+    AppendVariableInt((0x80 | len_num_bytes), len_pos);
+    len_num_bytes += 1;
+  }
+  return len_num_bytes;
+}
+
+size_t ASN1ProtoToDER::EncodeLength(const Length& len,
+                                    const size_t actual_len,
+                                    const size_t len_pos) {
+  if (len.has_length_override()) {
+    return EncodeOverrideLength(len.length_override(), len_pos);
+  } else if (len.has_indefinite_form() && len.indefinite_form()) {
+    return EncodeIndefiniteLength(len_pos);
+  } else {
+    return EncodeDefiniteLength(actual_len, len_pos);
+  }
+}
+
+size_t ASN1ProtoToDER::EncodeValue(const Value& val) {
+  size_t len = 0;
+  for (const auto& val_ele : val.val_array()) {
+    if (val_ele.has_pdu()) {
+      len += EncodePDU(val_ele.pdu());
+    } else {
+      len += val_ele.val_bits().size();
+      encoder_.insert(encoder_.end(), val_ele.val_bits().begin(),
+                      val_ele.val_bits().end());
+    }
+  }
+  return len;
+}
+
+uint8_t ASN1ProtoToDER::EncodeHighTagNumberForm(const uint8_t id_class,
+                                                const uint8_t encoding,
+                                                const uint32_t tag_num) {
+  // The high-tag-number form base 128 encodes |tag_num| (X.690 (2015), 8.1.2).
+  uint8_t num_bytes = GetNumBytes(tag_num, 128);
+  // High-tag-number form requires the lower 5 bits of the identifier to be set
+  // to 1 (X.690 (2015), 8.1.2.4.1).
+  uint64_t id_parsed = (id_class | encoding | 0x1F);
+  id_parsed <<= 8;
+  for (uint8_t i = num_bytes - 1; i != 0; --i) {
+    // If it's not the last byte, the high bit is set to 1 (X.690
+    // (2015), 8.1.2.4.2).
+    id_parsed |= ((0x01 << 7) | ((tag_num >> (i * 7)) & 0x7F));
+    id_parsed <<= 8;
+  }
+  id_parsed |= (tag_num & 0x7F);
+  AppendVariableInt(id_parsed, encoder_.size());
+  return num_bytes + 1;
+}
+
+uint8_t ASN1ProtoToDER::EncodeIdentifier(const Identifier& id) {
+  // The class comprises the 7th and 8th bit of the identifier (X.690
+  // (2015), 8.1.2).
+  uint8_t id_class = static_cast<uint8_t>(id.id_class()) << 6;
+  // The encoding comprises the 6th bit of the identifier (X.690 (2015), 8.1.2).
+  uint8_t encoding = static_cast<uint8_t>(id.encoding()) << 5;
+
+  uint32_t tag_num = id.tag_num().has_high_tag_num()
+                         ? id.tag_num().high_tag_num()
+                         : id.tag_num().low_tag_num();
+  // When the tag number is greater than or equal to 31, encode with a single
+  // byte; otherwise, use the high-tag-number form (X.690 (2015), 8.1.2).
+  if (tag_num >= 31) {
+    return EncodeHighTagNumberForm(id_class, encoding, tag_num);
+  }
+  AppendVariableInt((id_class | encoding | tag_num), encoder_.size());
+  return 1;  // low-tag-number form requires 1 byte to encode.
+}
+
+size_t ASN1ProtoToDER::EncodePDU(const PDU& pdu) {
+  ++depth_;
+  // Artifically limit the stack depth to avoid stack overflow.
+  if (depth_ > 67000) {
```

Comment:
How'd you come up with this number, fwiw? I think something like 200 is probably plenty sufficient?

```suggestion
  if (depth_ > kRecursionLimit) {
```

Reply:
Just by forcing a bunch of recursive calls and seeing how far we can go, then cutting off the number by a constant to be safe. But yeah you're right, we don't need that many! Change made.